src/_pytest/mark/structures.py

```diff
@@ -283,28 +284,37 @@ def test_function():
     on the ``test_function`` object. """
 
     _config = None
+    _markers = set()
 
     def __getattr__(self, name):
         if name[0] == "_":
             raise AttributeError("Marker name must NOT start with underscore")
+
         if self._config is not None:
-            self._check(name)
+            self._update_markers(name)
```

Comment:
Shouldn't this be:

```python
if self._config is not None:
    if name not in self._markers:
        self._update_markers(name) 
    if name not in self._markers:
        # warning or error
```

As it stands, we are updating the known marks every time a new mark is decorated.

Reply:
The `update_markers` method would only do anything if `name` was unknown, but point taken that this was confusing - I've inlined it into `__getattr__` to clarify the logic.

Reply:
Oh you are right. But I agree that better be more explicit, thanks.